version: "3.9"
services:
  postgres:
    container_name: databackend-db
    image: postgres:latest
    environment:
      POSTGRES_DB: databackend
      POSTGRES_USER: databackend
      POSTGRES_PASSWORD: databackend
      PGDATA: /var/lib/postgresql/data
    healthcheck:
      test: ['CMD-SHELL', 'pg_isready -U databackend']
      interval: 5s
      timeout: 60s
      retries: 30
    volumes:
      - databackend-db:/var/lib/postgresql/data
    ports:
      - "5433:5432"
    networks:
      - backend
    restart: unless-stopped
  
  analytics:
    container_name: analytics-db
    image: postgres:latest
    environment:
      POSTGRES_DB: analytics
      POSTGRES_USER: analytics
      POSTGRES_PASSWORD: analytics
      PGDATA: /var/lib/postgresql/data
    healthcheck:
      test: ['CMD-SHELL', 'pg_isready -U analytics']
      interval: 5s
      timeout: 60s
      retries: 30
    volumes:
      - analytics-db:/var/lib/postgresql/data
    ports:
      - "5434:5432"
    networks:
      - backend
    restart: unless-stopped

  sae:
    container_name: sae-db
    image:   timescale/timescaledb-ha:pg14-latest
    environment:
      POSTGRES_DB: sae
      POSTGRES_USER: sae
      POSTGRES_PASSWORD: sae
    healthcheck:
      test: ['CMD-SHELL', 'pg_isready -U sae']
      interval: 5s
      timeout: 60s
      retries: 30
    volumes:
      - ./data/init.sql:/docker-entrypoint-initdb.d/init.sql
      - sae-db:/home/postgres/pgdata/data
    ports:
      - "5435:5432"
    networks:
      - backend
    restart: unless-stopped

  pgadmin:
    container_name: pgadmin_container
    image: dpage/pgadmin4
    environment:
      PGADMIN_DEFAULT_EMAIL: pgadmin4@pgadmin.org
      PGADMIN_DEFAULT_PASSWORD: admin
      PGADMIN_CONFIG_SERVER_MODE: 'False'
    volumes:
       - databackend-pgadmin:/var/lib/pgadmin
    ports:
      - "5050:80"
    networks:
      - backend
    restart: unless-stopped

  databackend-db-keycloak:
    image: postgres:latest
    restart: on-failure
    environment:
      POSTGRES_DB: 'keycloak'
      POSTGRES_USER: 'keycloak'
      POSTGRES_PASSWORD: 'keycloak'
      PGDATA: /var/lib/postgresql/data
    healthcheck:
      test: ['CMD-SHELL', 'pg_isready -U keycloak'] 
      interval: 5s
      timeout: 60s
      retries: 30
    volumes:
      - databackend-keycloak-db:/var/lib/postgresql/data
    networks:
      - backend

  databackend-keycloak:
    image: quay.io/keycloak/keycloak
    volumes:
      - ./keycloak/imports:/opt/keycloak/data/import
    depends_on:
      databackend-db-keycloak:
        condition: service_healthy
    restart: on-failure
    environment:
      KC_DB_URL: jdbc:postgresql://databackend-db-keycloak:5432/keycloak
      KC_DB: postgres
      KC_DB_USERNAME: keycloak
      KC_DB_PASSWORD: keycloak
      KEYCLOAK_ADMIN: admin
      KEYCLOAK_ADMIN_PASSWORD: admin
      KC_HTTP_RELATIVE_PATH: /auth/
    command: start-dev --import-realm
    ports:
      - '8080:8080'
    networks:
      - backend


networks:
  backend:

volumes:
  databackend-db:
  databackend-pgadmin:
  databackend-keycloak-db:
  analytics-db:
  sae-db:

